### PR TITLE
Time - Yesenia

### PR DESCRIPTION
**Pull request by yesentorres**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | One custom model method I wrote is for the Work model and it is called sort_works. Its purpose is to create an array of all works in the database labeled with the category specified in the method argument. If there are no works with that category, it returns a string saying that there are no works for that category. If there are works, it returns an array of those categorized works in descending order based on vote count.
Describe how you approached testing that model method. What edge cases did you come up with? | The edge case I tested was the case when there are no works. I tested to make sure the array came back in descending order when there were works. 
What are `session` and `flash`? What is the difference between them? | Session and flash are hashes provided by Rails that helps you to persist data. Session lets you persist data over multiple request-response cycles, whereas flash persists data for one request cycle.
What was one thing that you gained more clarity on through this assignment? | I gained more clarity on the MVC model and how to apply within an actual app. 
What is the Heroku URL of your deployed application? | https://yt-media-ranker.herokuapp.com/

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+Rails.application.routes.draw do
```
Comment (CheezItMan) [inline config/routes.rb:1]: Very nicely done here

File: test/controllers/users_controller_test.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+describe UsersController do
```
Comment (CheezItMan) [inline test/controllers/users_controller_test.rb:3]: Just noting no tests here

File: test/controllers/works_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+require "test_helper"
+
+describe WorksController do
```
Comment (CheezItMan) [inline test/controllers/works_controller_test.rb:3]: Just noting no tests here

File: app/models/work.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
+class Work < ApplicationRecord
+  validates :title, presence: true
+  validates_uniqueness_of :title, scope: :category
+
+  has_many :votes, dependent: :destroy
+
+  def self.sort_works(category)
```
Comment (CheezItMan) [inline app/models/work.rb:7]: Suggestion, make the valid categories a constant array and then you can check to see if the category is valid with CATEGORIES.include? category

Then you can use the variable with `self.where(category: category)`

File: app/controllers/works_controller.rb [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    @works = Work.all
+  end 
+
+  def index
+    @works = Work.all
+  end
+
+  def show
+    if @work.nil?
+      render_not_found
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:15]: Great!

File: app/controllers/works_controller.rb [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  def create
+    @work = Work.new(work_params) 
+
+    if @work.save
+      flash[:success] = "Successfully created #{@work.category} #{@work.id}"
+      redirect_to work_path(@work.id)
+      return
+    else 
+      flash.now[:error] = "A problem occured: Could not create #{@work.category}"
+      render :new
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:33]:
```suggestion
      render :new, status: :bad_request
```

File: app/controllers/works_controller.rb [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    if @work.nil?
+      render_not_found
+      return
+    elsif @work.update(work_params)
+      flash[:success] = "Successfully updated #{@work.category} #{@work.id}"
+      redirect_to work_path(@work.id)
+      return
+    else 
+      flash.now[:error] = "A problem occured: Could not update #{@work.category}"
+      render :edit
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:55]:
```suggestion
      render :edit, status: :bad_request
```

File: app/controllers/works_controller.rb [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+      if @vote.save
+        flash[:success] = "Successfully upvoted!"
+        redirect_back fallback_location: work_path(@work)
+        return 
+      else 
+        flash.now[:error] = "A problem occured: Could not upvote"
+        # create instance variable to pass the vote errors to the views
+        @invalid_vote_message = @vote.errors.messages
+
+        render :show
```
Comment (CheezItMan) [inline app/controllers/works_controller.rb:89]:
```suggestion
        render :show, status: :bad_request
```

File: app/views/works/main.html.erb [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        </ul>
+      <% end %>
+    </section>
+
+    <section class="top-ten__list-container">
+      <h2 class="top-ten__header"> Top Books </h2>
+      <% if books.instance_of? String %>
+        <p><%= books %></p>
+      <% else %>
+        <ul class="list-group top-ten__list">
```
Comment (CheezItMan) [inline app/views/works/main.html.erb:69]: This part can be done with a partial view.

**Comment by CheezItMan** [comment]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ⚠ No filter to find a user
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️, fixtures for users and works, but not votes.
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️






## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅


## Summary


Nicely done, you hit all the learning goals here.  You don't have controller tests, but otherwise this is outstanding.  Take a look at my comments and let me know what questions you have.
